### ISSUE TITLE
Look into data type optimization in socket transfers

### ISSUE THREAD

**Issue by hclivess** [hclivess/Bismuth#172]
sys.getsizeof() returns smaller size on proper data types (list), however, sockets.py converts everything to/from strings (remainder of former serialization). Look into the possibility of dropping strings are reducing data transfer size in multitudes.

For example:

```
[[[1515443350.96, 'd9dcbda8025fd296d842289bba7a1bd2fab5c63decf73d0458bbea70', 'd9dcbda8025fd296d842289bba7a1bd2fab5c63decf73d0458bbea70', 0, 'I/SOMu99A0ltj9LHVJltza8iBy4vFni3RkCW7TLiD8wj5QEooMagQM59kxZms/4tSMAIRoB70TnXX/dZhj4rIVW2km2Tcx/iN9R6fqBv1KdcgMmhNONJeP72nDuX1tcIPSrTYwM44CzAnjfRBOFfbjHI17PvzoAmAenm9Oy+wtBh8elk8C3LuDV/KaOxBHZyFH/vVQ9yqqTgfmGcAaI5IPB8TJX+pFtXYkz867iM6SgNO+d+2G2qMG/73mxAVVciYmFnOx8hZMDLlAeZr7k/Gxy8onMEvJHfDsjHKi+8sl/33B+Gz7VZ+H8x8ocLeqrzgRL9XXh7iJbq9ekUQH29dLVhb08fZnjKhd1sNxpjukKGKC6cLCVvn4is3GV/sVYiDQBV6HjJpTFGQnylIcJJHQCBw8Szlud52q3cj4e79r+imUal8D8HsoYsPB9YCj2Ke7xZmI1ix4F18Pmn178hWuhSerc61zwFlbe6w+Yk8AL98Mc1rDtLhj55MZNhuLndnQmCM98+Pe3z6CofptNYqwqwwdN6/7Zq9FwgHO5asvSBYFCLJzUOsk90eygy7Xje+PTkqnn7oyAIAZ6YjqJ/uGVUuGBFvC6jmzGXJDaeuF7VB9nRzTTbffcyZrPTRt8/wKcRjyPeBZOHG79eutXmz8IOxEIOuoCZjCMz4+uApno=', 'LS0tLS1CRUdJTiBQVUJMSUMgS0VZLS0tLS0KTUlJQ0lqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FnOEFNSUlDQ2dLQ0FnRUFuVmJIQmVldWY3U1o3RlZLeFhLSgovcnNwaTBNVllyQkRqMFJNZFlQbW11UDFLbEJCQmtKMUJ6THUxK1ZjeGtwaGZrOWdNZU5ra3JhWFdwVzJqa3pPCnlybHdvZFAyNWZuUzF4bThFNEsxa3Rib0laTXJFWFpjS1UyRGVTbVV6elo1NldrSWNQQWNtZEVYVzhXVi9YQnAKekRwM1gwWnVacjk1Y3Vtd0RZcUQwWE9DTnB4QUpnWVB1Y3V3SjNmaXUvKzE3MGp1NHU5YlZZSnRiYS9uUDJNbgpqMWxaZGdmbjN4Yzl6VThTeWZmYi9oNWh1T082cW5sT2VFbTRGSXZJSzByeDhVMmJVSnZYeU81RjNJalIvV0x2CkJzMm5LWlJ6ZHNTRFFqcGUwUVZ0RzVzTnBWc2QrRDNWTFliQzFjRGNMQmJWalFrVEVZTlFOVXhZdFF5QnNRMzIKbUhLdkdIVVFxaXdXTEVxUTFOay9haFhBWW9oTzlweXY2djZTeEE3OFZ5YkJYbTBDVkx4VGRETlhKK3kvajBLbwpjSzcxTUQ5NlRLWTZRMjFCNmZrL3hlbGhMMmIyNm1IU0JlalFXdW1BV2hSRmJqWHhXZkhaYWE2c3NpbWVLNEZxCkVzMS8veFR5OVMrQVluaVk1YkU2YmEvK2RvOUd6Wm5sODdVS2NCL2V3RDA0SXhtRmkzUDJQcWZnQnN4V2ZsMFMKdkd0Rll1TC9iNVZFSzNtcldXdEp6QXhaT2g5L0tXNUpXK0g1OXgzVmI4Q2NzU3luR09ldUNMSnJkSDB2am9LNQpOdnFPNXlCMTdVVkxrcjNjdlhiWEVXWGlsUmVoc2V5dEhWcURLQTUrTmVNc1JjOGlqd1E2TVpoWFRHTXBsMGVJCjN4UndJaE4rVkpOTmRxZXczc3A2cGE4Q0F3RUFBUT09Ci0tLS0tRU5EIFBVQkxJQyBLRVktLS0tLQ==', '0', '608ea1cc0000000010ba7dd801010000']]]


<class 'str'>
1992
Block size: 0.001992 MB

<class 'list'>
72
Block size: 0.000072 MB

```
That's a reduction of a 27.6x

**Comment by hclivess** [comment]
related: https://stackoverflow.com/questions/23311221/sys-getsizeof1-and-sys-getsizeof10000-return-the-same-output-why

**Comment by hclivess** [comment]
Also summoning @Aareon and @HarryR  for a possible comment on this

**Comment by hclivess** [comment]
@EggPool will crate a PoC on https://developers.google.com/protocol-buffers/docs/overview

**Comment by Aareon** [comment]
```py
#python 3.5.2

from sys import getsizeof

a = [1515443350.96, 'd9dcbda8025fd296d842289bba7a1bd2fab5c63decf73d0458bbea70', 'd9dcbda8025fd296d842289bba7a1bd2fab5c63decf73d0458bbea70', 0, 'I/SOMu99A0ltj9LHVJltza8iBy4vFni3RkCW7TLiD8wj5QEooMagQM59kxZms/4tSMAIRoB70TnXX/dZhj4rIVW2km2Tcx/iN9R6fqBv1KdcgMmhNONJeP72nDuX1tcIPSrTYwM44CzAnjfRBOFfbjHI17PvzoAmAenm9Oy+wtBh8elk8C3LuDV/KaOxBHZyFH/vVQ9yqqTgfmGcAaI5IPB8TJX+pFtXYkz867iM6SgNO+d+2G2qMG/73mxAVVciYmFnOx8hZMDLlAeZr7k/Gxy8onMEvJHfDsjHKi+8sl/33B+Gz7VZ+H8x8ocLeqrzgRL9XXh7iJbq9ekUQH29dLVhb08fZnjKhd1sNxpjukKGKC6cLCVvn4is3GV/sVYiDQBV6HjJpTFGQnylIcJJHQCBw8Szlud52q3cj4e79r+imUal8D8HsoYsPB9YCj2Ke7xZmI1ix4F18Pmn178hWuhSerc61zwFlbe6w+Yk8AL98Mc1rDtLhj55MZNhuLndnQmCM98+Pe3z6CofptNYqwqwwdN6/7Zq9FwgHO5asvSBYFCLJzUOsk90eygy7Xje+PTkqnn7oyAIAZ6YjqJ/uGVUuGBFvC6jmzGXJDaeuF7VB9nRzTTbffcyZrPTRt8/wKcRjyPeBZOHG79eutXmz8IOxEIOuoCZjCMz4+uApno=', 'LS0tLS1CRUdJTiBQVUJMSUMgS0VZLS0tLS0KTUlJQ0lqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FnOEFNSUlDQ2dLQ0FnRUFuVmJIQmVldWY3U1o3RlZLeFhLSgovcnNwaTBNVllyQkRqMFJNZFlQbW11UDFLbEJCQmtKMUJ6THUxK1ZjeGtwaGZrOWdNZU5ra3JhWFdwVzJqa3pPCnlybHdvZFAyNWZuUzF4bThFNEsxa3Rib0laTXJFWFpjS1UyRGVTbVV6elo1NldrSWNQQWNtZEVYVzhXVi9YQnAKekRwM1gwWnVacjk1Y3Vtd0RZcUQwWE9DTnB4QUpnWVB1Y3V3SjNmaXUvKzE3MGp1NHU5YlZZSnRiYS9uUDJNbgpqMWxaZGdmbjN4Yzl6VThTeWZmYi9oNWh1T082cW5sT2VFbTRGSXZJSzByeDhVMmJVSnZYeU81RjNJalIvV0x2CkJzMm5LWlJ6ZHNTRFFqcGUwUVZ0RzVzTnBWc2QrRDNWTFliQzFjRGNMQmJWalFrVEVZTlFOVXhZdFF5QnNRMzIKbUhLdkdIVVFxaXdXTEVxUTFOay9haFhBWW9oTzlweXY2djZTeEE3OFZ5YkJYbTBDVkx4VGRETlhKK3kvajBLbwpjSzcxTUQ5NlRLWTZRMjFCNmZrL3hlbGhMMmIyNm1IU0JlalFXdW1BV2hSRmJqWHhXZkhaYWE2c3NpbWVLNEZxCkVzMS8veFR5OVMrQVluaVk1YkU2YmEvK2RvOUd6Wm5sODdVS2NCL2V3RDA0SXhtRmkzUDJQcWZnQnN4V2ZsMFMKdkd0Rll1TC9iNVZFSzNtcldXdEp6QXhaT2g5L0tXNUpXK0g1OXgzVmI4Q2NzU3luR09ldUNMSnJkSDB2am9LNQpOdnFPNXlCMTdVVkxrcjNjdlhiWEVXWGlsUmVoc2V5dEhWcURLQTUrTmVNc1JjOGlqd1E2TVpoWFRHTXBsMGVJCjN4UndJaE4rVkpOTmRxZXczc3A2cGE4Q0F3RUFBUT09Ci0tLS0tRU5EIFBVQkxJQyBLRVktLS0tLQ==', '0', '608ea1cc0000000010ba7dd801010000']

b = str(a)
print(type(b))
size = getsizeof(b)
print(size)
print("Block size: {} MB\n".format(size/10000))

print(type(a))
size = getsizeof(a)
print(size)
print("Block size: {} MB\n".format(size/10000))

c = set(a)
print(type(c))
size = getsizeof(c)
print(size)
print("Block size: {} MB\n".format(size/10000))

d = tuple(a)
print(type(d))
size = getsizeof(d)
print(size)
print("Block size: {} MB\n".format(size/10000))
```

**Comment by HarryR** [comment]
Some recommendations on reducing the size:

 * Don't hex or base64 encode data, use raw bytes
 * Don't use RSA, use elliptic curve crypto (e.g. ed25519, secp256k1) - where the public key is 32 bytes

**Comment by Aareon** [comment]
These are awesome suggestions

**Comment by HarryR** [comment]
With a record format like:

```python
BismuthTxn = namedtuple('BismuthTxn', ('timestamp', 'from', 'to', 'value', 'extradata' 'signature'))
```

 * The `from` and `to` fields are 32 byte compressed ECC public keys.
 * `signature` is a 64 byte Schnorr signature.
 * `timestamp` is a 4 byte integer
 * `extradata` is an arbitrary chunk of bytes

This makes the record size `132 + len(extradata)` bytes. Storing it as a `namedtuple` reduces the memory overhead too.

**Comment by EggPool** [comment]
Good suggestions!

Working on protobuff, for internal storage as well as socket transmission.
Adresses and hashes are then stored as bytes, not hex, yes.
msgpack is indeed better than Json. Protobuff looks even smaller and also provides validity checks and no cost (size speaking) optional properties.
I may try a poc variant with msgpack, to compare speed/size.

Elliptic curve would make it difficult to keep compatibility with existing addresses I think.
Some things will be left sub-optimal in order to keep compatibility, I'm afraid.

**Comment by HarryR** [comment]
It should be possible to support both RSA and ECC keys / signatures.

e.g. replace the RSA public key field with an algorithm specifier, e.g. 'p256k1'

**Comment by Aareon** [comment]
@EggPool @hclivess perhaps I’m misunderstanding, but is there a reason the example block above is encapsulated in 3 brackets on either side? Why would it need to be a nested list within a nested list within a nested list?

**Comment by Aareon** [comment]
At most I can see this being a nested list within a nested list, but not what’s above. Even then, it’s a single block. Just trying to understand.

**Comment by HarryR** [comment]
From what I recall (given my hash-addled and somewhat autistic mind), there is another problem related to the serialisation of Python objects to strings...

Check: https://github.com/hclivess/Bismuth/blob/master/node.py#L1296

In order to verify the transaction a hash is made using a Python encoded string of the arguments, the arguments are a tuple of things and the whole thing is converted to a string, this binds everything to the Python encoding format.

I have many more qualms with the Bismuth codebase, and no offence but it could do with some objective refactoring and serious software engineering given that it's apparently worth a bunch of imaginary foreign coins on a couple of OTC exchanges.

Here is some code I wrote, because lol:

```
#!/usr/bin/env python
"""
# Minimal Plasma Chain

This is a minimal 'plasma chain', it's a fast payment network 
which can be used to create and exchange tokens across multiple
block-chains.


# TODO:

 * Payment dependencies
 * If dependency doesn't exist, the block payment is dropped
 * Does this add too much complexity?


## Processing Payments

Each payment contains the following information:

 from - address: source
 to - address: destination
 currency - address: currency
 value - int256: signed integer
 ref - bytes32: unique payment reference
 sig - ecdsa: recoverable signature

The signed message (and transaction ID) is the following:

 H(prev-blockhash || from || to || currency || value || ref)

Because each signed payment is only valid for one-block it must 
be re-signed and re-broadcast until the payment is accepted by
the network.

### Minting Tokens

Anybody is allowed to create as many tokens of their own currency
as they want by sending a special 'mint' transaction from itself to
itself using its own address as the currency.


## Smart-contract Integration

If an on-chain contract uses a trusted merkle root relay it can verify if
a payment has been made to it. After sending a payment to a contract address
an on-chain transaction must activate the contract and supply it with the
merkle proof.

The merkle proof ensures:

 - The `value` of `currency` (the contracts own address) has been
   permanently removed from the senders balance.

 - There is opportunity to use a unique reference to prevent
   double-spend by the contract.

The contract only needs the `value` and `ref` parameters and the
merkle proof (block height height, path), it has access to `msg.sender`
and its own address which is everything necessary to validate a payment
exists in the merkle root of a trusted side-chain.

The blockchain is monitored for all plasma compatible events from contracts
which are then recorded just as regular payments would be.


## Block Format

Blocks contain three parameters:

    hash - bytes32 - H(id, root, prev-hash)
    id - integer - Block sequence number
    root - bytes32 - Merkle tree root


## Merkle Proofs

Each block is a merkle tree which contains three types of items:

 - payments
 - signatures
 - balances

### Record types

Payment entry format is 124 bytes:

  - from - address
  - to - address
  - currency - address
  - value - bytes32
  - ref - bytes32

Signature format is 65 bytes:

  - v - byte
  - r - uint256
  - s - uint256

Balance format is 72 bytes:

  - currency - address
  - account - address
  - balance - uint256


# References

 * http://cryptonite.info/files/mbc-scheme-rev3.pdf
"""


from __future__ import print_function
import os
from base64 import b64encode, b64decode

try:
    import cPickle as pickle
except ImportError:
    import pickle
import argparse
from collections import namedtuple, defaultdict

from ethereum.utils import sha3, zpad, int_to_big_endian, privtoaddr, int_to_32bytearray, big_endian_to_int, encode_int32
from rlp.utils import ascii_chr

try:
    import coincurve
except ImportError:
    from py_ecc.secp256k1 import ecdsa_raw_recover, ecdsa_raw_sign
    import warnings
    warnings.warn('could not import coincurve', ImportWarning)
    coincurve = None

from merkle import *

if sys.version_info.major == 2:
    def bytearray_to_bytestr(value):
        return bytes(''.join(chr(c) for c in value))
else:
    def bytearray_to_bytestr(value):
        return bytes(value)


u256be = lambda v: zpad(int_to_big_endian(v), 32)

flatten = lambda l: [item for sublist in l for item in sublist]

# Turns a `defaultdict(defaultdict)` into a flat dictionary
dict_dump = lambda diff: {c: dict(d.items()) for c, d in diff.items()}

def require(*args):
    if not all(args):
        raise RuntimeError("Requirement failed")


# --------------------------------------------------------------------
# Types


Block = namedtuple('Block', ('hash', 'id', 'root'))
Payment = namedtuple('Payment', ('f', 't', 'c', 'v', 'r'))
EcdsaSignature = namedtuple('EcdsaSignature', ('v', 'r', 's'))
SignedPayment = namedtuple('SignedPayment', ('p', 's'))

def marshal(x):
    if isinstance(x, (int, long)):
        return x
    if isinstance(x, (str, bytes, unicode)):
        return b64encode(x)
    raise ValueError("Cannot marshal type: %r - %r" % (type(x), x))

def unmarshal(x):
    if isinstance(x, (int, long)):
        return x
    if isinstance(x, (str, bytes, unicode)):
        return b64decode(x)
    raise ValueError("Cannot unmarshal type: %r - %r" % (type(x), x))


def Block_marshal(b):
    return Block(*map(marshal, b))

def Block_unmarshal(b):
    return Block(*map(unmarshal, b))

def Payment_marshal(p):
    return Payment(*map(marshal, p))

def Payment_unmarshal(p):
    return Payment(*map(unmarshal, p))

def EcdsaSignature_marshal(sig):
    return EcdsaSignature(*map(marshal, sig))

def EcdsaSignature_unmarshal(sig):
    return EcdsaSignature(*map(unmarshal, sig))

def SignedPayment_marshal(sp):
    return SignedPayment(Payment_marshal(sp[0]), EcdsaSignature_marshal(sp[1]))

def SignedPayment_unmarshal(sp):
    return SignedPayment(Payment_unmarshal(sp[0]), EcdsaSignature_unmarshal(sp[1]))


# --------------------------------------------------------------------
# Data persistence load/save


data_load = lambda x: pickle.load(open(x))
data_save = lambda d, x: pickle.dump(d, open(x, 'wb'))

chaindata_file = lambda n, x: "%08X.%s.pickle" % (n, x,)
chaindata_path = lambda n, x: "chaindata/" + chaindata_file(n, x)
chaindata_load = lambda n, x: data_load(chaindata_path(n, x))
chaindata_save = lambda n, x, d: data_save(d, chaindata_path(n, x))
chaindata_exists = lambda n, x: os.path.exists(chaindata_path(n, x))

block_exists = lambda n: chaindata_exists(n, 'block')
block_save = lambda n, d: chaindata_save(n, 'block', d)
block_load = lambda n: chaindata_load(n, 'block')
diff_save = lambda n, d: chaindata_save(n, 'diff', d)
payments_save = lambda n, d: chaindata_save(n, 'payments', d)
balances_save = lambda n, d: chaindata_save(n, 'balances', d)
balances_load = lambda n: chaindata_load(n, 'balances')


def chaindata_latest_set(n):
    """Maintain symlink to latest block"""
    latest_path = chaindata_path(0, 'latest')
    if os.path.exists(latest_path):
        os.unlink(latest_path)
    os.symlink(chaindata_file(n, 'block'), latest_path)


def chaindata_latest_get():
    """Which is the most recent block"""
    latest_path = chaindata_path(0, 'latest')
    if os.path.exists(latest_path):
        block = data_load(latest_path)
        return block.id


# --------------------------------------------------------------------


# TODO: manage a block of payments
#     - check if all payments balance etc.
#     - add / remove individual payments from the block
#     - check dependencies between payments, evicting dependent payments
#     - given a block, determine payment status
#     - verify payment reference doesn't already exist


def block_genesis():
    if not os.path.exists('chaindata'):
        os.mkdir('chaindata')
    block_save(0, Block(hash=os.urandom(32), id=0, root=os.urandom(32)))
    payments_save(0, list())
    balances_save(0, dict())
    chaindata_latest_set(0)


# TODO: atomic bilateral swap


def payments_apply(payments):
    # Calculate resulting balance in each currency for every party
    balances = defaultdict(lambda: defaultdict(int))
    created = defaultdict(int)
    for p in payments:
        if p.v <= 0:
            raise RuntimeError("Payment value < 0: %r" % (p,))
        if p.f == p.c and p.f == p.t:
            # Self transaction mints your own currency
            created[p.f] += p.v
        else:
            balances[p.c][p.f] -= p.v
        balances[p.c][p.t] += p.v
    return balances, created


def diff_apply(diff, balances):
    for c, d in diff.items():
        for k, v in d.items():
            if c not in balances:
                balances[c] = dict()
            if k not in balances[c]:
                balances[c][k] = 0
            balances[c][k] += v
    return balances


def diff_dump(diff, balances):
    """Provides proof of balance in packed format:
         currency || address || balance"""
    return [''.join([zpad(c, 20), zpad(k, 20), u256be(balances.get(c, dict()).get(k, 0) + v)])
            for c, d in diff.items()
            for k, v in d.items()]


def diff_is_balanced(diff, created):
    """Check if the sum value for all payments for each currency sum to zero"""
    return all([ sum(d.values()) - created[c] >= 0
                 for c, d in diff.items() ])


def diff_is_funded(diff, created, balances):
    """If, after applying balance diff, are all balances above zero?"""
    for c, d in diff.items():
        for k, v in d.items():
            balance = balances.get(c, dict()).get(k, 0) + v
            if c == k:
                balance += created.get(k, 0)
            if balance < 0:
                return False
    return True


def payment_dump(p):
    require( all([len(x) == 20 for x in [p.f, p.t, p.c]]) )
    require( len(p.r) == 32 )
    require( isinstance(p.v, (int, long)) )
    return ''.join([ p.f, p.t, p.c, u256be(p.v), p.r ])


payment_hash = lambda payment, prev_hash: sha3(prev_hash + payment_dump(payment))


def ecsign(rawhash, key):
    if coincurve and hasattr(coincurve, 'PrivateKey'):
        pk = coincurve.PrivateKey(key)
        signature = pk.sign_recoverable(rawhash, hasher=None)
        v = safe_ord(signature[64]) + 27
        r = big_endian_to_int(signature[0:32])
        s = big_endian_to_int(signature[32:64])
    else:
        v, r, s = ecdsa_raw_sign(rawhash, key)
    return v, r, s


def payment_seal(payment, secret, prev_hash):
    require( len(secret) == 32 )
    msg = payment_hash(payment, prev_hash)
    sig = ecsign(msg, secret)
    return SignedPayment(payment, EcdsaSignature(*sig))


def signature_recover(rawhash, sig):
    v, r, s = sig
    if coincurve and hasattr(coincurve, "PublicKey"):
        try:
            pk = coincurve.PublicKey.from_signature_and_message(
                ''.join([ zpad(bytearray_to_bytestr(int_to_32bytearray(r)), 32),
                          zpad(bytearray_to_bytestr(int_to_32bytearray(s)), 32),
                          ascii_chr(v - 27) ]),
                rawhash,
                hasher=None,
            )
            pub = pk.format(compressed=False)[1:]
        except BaseException:
            pub = b"\x00" * 64
    else:
        result = ecdsa_raw_recover(rawhash, (v, r, s))
        if result:
            x, y = result
            pub = encode_int32(x) + encode_int32(y)
        else:
            raise ValueError('Invalid VRS')
    assert len(pub) == 64
    return sha3(pub)[12:]


signature_dump = lambda s: ''.join([chr(s.v), u256be(s.r), u256be(s.s)])


def payments_sum(payments):
    outv = defaultdict(int)
    outr = defaultdict(list)
    for p in payments:
        outv[(p.f,p.t,p.c)] += p.v
        outr[(p.f,p.t,p.c)] += p.r
    return [ Payment(f, t, c, v, outr[(f,t)])
             for (f, t, c), v in outv.items() ]


def signedpayment_verify(sp, prev_hash):
    msg = payment_hash(sp.p, prev_hash)
    f = signature_recover(msg, sp.s)
    return f == sp.p.f


def signedpayment_open(sp, prev_hash):
    if not signedpayment_verify(sp, prev_hash):
        raise ValueError("Invalid signature,")
    return sp.p


def block_seal(id, signed_payments, balances, prev_hash):
    """
    Seal block by applying signed payments to balances
    """
    require( len(prev_hash) == 32 )

    # Verify that payments are valid at the current state
    payments = map(lambda x: signedpayment_open(x, prev_hash), signed_payments)
    balances_diff, created = payments_apply(payments)

    # TODO: check for duplicate payment references...

    require( diff_is_balanced(balances_diff, created), "Misbalance" )
    require( diff_is_funded(balances_diff, created, balances), "Not funded" )

    # Store payments, signatures and balances in merkle tree
    dumped_sigs = sorted(map(signature_dump, [sp.s for sp in signed_payments]))
    dumped_payments = sorted(map(payment_dump, payments))
    dumped_balances = sorted(diff_dump(balances_diff, balances))
    tree, root = merkle_tree(dumped_sigs + dumped_payments + dumped_balances)

    block_hash = sha3(prev_hash + ''.join(map(u256be, [
        id, root
    ])))
    return Block(block_hash, id, root), balances_diff


def blockchain_apply(prev_block_no, signed_payments):
    prev_block = block_load(prev_block_no)
    block_no = prev_block_no + 1
    if block_exists(block_no):
        raise RuntimeError("Cannot overwrite block", block_no)

    prev_balances = balances_load(prev_block_no)
    block, balances_diff = block_seal(block_no, signed_payments, prev_balances, prev_block.hash)

    block_save(block_no, block)
    chaindata_save(block_no, 'diff', dict_dump(balances_diff))
    payments_save(block_no, signed_payments)
    payments_graphviz([sp.p for sp in signed_payments]).render('chaindata/%08X.graphviz' % (block_no,))    

    after_balances = diff_apply(balances_diff, prev_balances)
    balances_save(block_no, after_balances)
    chaindata_latest_set(block_no)
    return block


# --------------------------------------------------------------------


def random_keypairs(n):
    keys = [os.urandom(32) for _ in range(0, n)]
    pubs = map(privtoaddr, keys)
    return dict(zip(pubs, keys))


def random_payment(balances, keypairs):
    # TODO: make code smaller, easier to read...
    while True:
        if len(balances) and random.choice([False, True]):
            c = random.choice(balances.keys())
            f, b = random.choice(balances[c].items())
            t = random.choice(keypairs.keys())
            if f == t:
                continue
        else:
            f = t = c = random.choice(keypairs.keys())
            b = 10000
        if b <= 0:
            continue
        v = random.randint(0, b)
        if v <= 0:
            continue
        if f != t or c != f:
            balances[c][f] -= v
        balances[c][t] += v
        return balances, Payment(f, t, c, v, os.urandom(32))


def random_payments(prev_hash, npayments=60, nparticipants=6):
    keypairs = random_keypairs(nparticipants)
    balances = defaultdict(lambda: defaultdict(int))
    out = []
    while len(out) < npayments:
        balances, p = random_payment(balances, keypairs)        
        out.append( payment_seal(p, keypairs[p.f], prev_hash) )
    return out


def payments_graphviz(payments, colours=dict()):
    payments = payments_sum(payments)
    import graphviz
    d = graphviz.Digraph(engine='circo')

    maxvals = defaultdict(list)
    for p in payments:
        maxvals[p.c].append(p.v)
    maxvals = {c: max(v) for c, v in maxvals.items()}

    for p in payments:
        d.edge(p.f[:4].encode('hex'), p.t[:4].encode('hex'),
               weight=str(p.v), fontsize="8.0",
               label=' '.join([str(p.v), p.c[:3].encode('hex')]),
               color=colours.get(p.c, 'black'),
               penwidth=str((float(abs(p.v)) / maxvals[p.c]) * 5))
    return d


# --------------------------------------------------------------------


def chain_options():
    parser = argparse.ArgumentParser(description="Plasma Chain")
    parser.add_argument('-n', '--block', dest='blockno', type=int,
                        help='Block number', default=0)
    parser.add_argument('-g', '--genesis', dest='genesis', action='store_true',
                        help="Create new genesis block")
    parser.add_argument('-r', '--random', dest='random', type=int, default=0,
                        help="Create a block of random payments")
    parser.add_argument('payments', nargs='*', help='Payments file')
    return parser.parse_args()


def main():
    opts = chain_options()
    if not chaindata_exists(0, 'block'):
        if not opts.genesis:
            raise ValueError("Must create genesis block")
        block_genesis()

    # TODO: find latest block number

    block = block_load(opts.blockno)

    # Create a block of random payments
    if opts.random:
        signed_payments = random_payments(block.hash, opts.random)
        block = blockchain_apply(opts.blockno, signed_payments)
        print("%d = %s\n" % (block.id, block.hash.encode('hex')))
        return 1

    # Otherwise, process payment files to create blocks
    for payment_file in opts.payments:
        signed_payments = data_load(payment_file)
        block = blockchain_apply(block.id, signed_payments)


if __name__ == "__main__":
    main()
````

**Comment by Aareon** [comment]
Very well documented and well formatted. This is an example of what the code base should aspire to, no offense to those that worked hard to make it. I agree entirely that some objective refactoring, reasonable compliance with PEP, and some more research will go a long way.

@hclivess in another issue, you said that you don't hold much respect for PEP-8. Is there a reason for this? No offense, but those proposals are there to help. Following them as reasonably close as possible can help make a large project clean, easily maintainable, and future-safe before it even becomes an issue. I'm just curious how any PEP could be a bad thing?

**Comment by Aareon** [comment]
As the project is experiencing now is an issue with modularity, reusability, and poor documentation, I feel like if only those same PEPs that we don't "hold much respect for" could have made the issues today much easier to correct.

**Comment by Aareon** [comment]
@HarryR Though, this is only just an example, I wonder if it would be too much for me to comment on a couple of things, if you don't mind?

**Comment by HarryR** [comment]
re pep-8, I have no respect for it either... Long story, but I prefer people point out where the code isn't readable and bitch about how they don't understand it, because that gains real insight IMO.

Also, please no need to give @hclivess a hard time, they're amazing for having the persistence to keep this thing going for as long as it has, and for that alone I have a huge amount of respect.

I have a bunch of other code that I'm trying to get to the MVP stage so it can be released, but it has its own internal priorities that have to be met first. But the difference between that, a handful of other 'python blockchain' projects and what @hclivess has made is the follow through and persistence.

**Comment by Aareon** [comment]
Oh, no, absolutely. I don't mean to give anyone a hard time. Like I said, props to them for being so persistent in working on this. I'm only trying to provide constructive criticism. As for myself, I prefer not to get messages from people telling me my code looks bad. I try to go extra lengths to make my code as readable and reusable as possible. That's just me.

I am by no means a professional, I'm only trying to help is all. Not trying to give anyone a hard time.

**Comment by Aareon** [comment]
Maybe I've been letting on a different vibe. I feel like everyone thinks I'm being condescending when I'm honest to God just trying to help. I apologize if it seems like I'm being rude.

**Comment by HarryR** [comment]
No worries, seriously :+1: 

There's a handful of things I've encountered with the Bismuth codebase which could do with improvement, and the worst thing I've learned about it is that the worse it gets the better it can get. For myself, personally, I've been a massive c*nt to a bunch of people and only realised later on how much knowledge I've gained in the process simply because at the time I didn't care.

I guess the perspective now is... so we have a blockchain, it's written in Python and we all know Python, we all have our own insights and ways to benefit the project... but how do we do this collaborative stuff - but a higher level question to that is - what is the real value of the project, other than it being easily approachable to Python programmers?

*shrugs*

I am working on distributed exchange methods, Ethereum and decentralisation. If I had one aim it would be to make sure Bismuth can be exchanged for anything without having to go through an exchange - we should hold the power, not them...

**Comment by Aareon** [comment]
And you guys are magnitudes more knowledgeable on cryptocurrency technology. I **really** want to contribute more and provide some massive contribution, but without a more extensive knowledge or the ability to constantly ask people questions, it's slow going.

I agree entirely that it is kind of exciting knowing that there is work to be done and being that there's already a working chain, it can only get better from here.

**Comment by HarryR** [comment]
Not really, I am cryptocurrency noob, I hack firmware for a living because Chinese programmers are worse than me ;)

I think the fundamental problem is an upgrade path.

1) We have a client-base, people who are mining, who are running the software

2) There are design choices which are non-optimal, with hindsight we can see that now

3) The blockchain is immutable, we have to stick with everything that has existed in the past, it has to be compatible with everything otherwise 'value' disappears...

Tbh the best thing that can happen right now is work out how to parse the blockchain, write a very very short chunk of Python which can validate the blockchain and ensure that everybody who has mined it keeps their value, and start to look at the smallest amount of code that can do the Bismuth 'core' bit...

Like, condensing all of the collective energy and wisdom that's gone into the project into the smallest amount of code possible, but that's a daunting task because it's probably split into a handful of segments:

 * The 'blockchain', validation of it
 * Transaction pool
 * p2p networking
 * tools & utilities
 * community needs

Like, you're on an exchange now - you cant ditch the project and switch to 'something new and more awesome' because you'll lose value.

Hmm.... tbh I see more value in a better more well designed blockchain which takes everything that Bismuth has accomplished, keeps the value that everybody has gained, but starts fresh at the same time, that's not an easy task though.

</harry... gone sleep now>

**Comment by hclivess** [comment]
String conversion is present because Python 2.x and changing the algorithm would require a lot of network cooperation, as noted above this comment. It does not threaten or hinder, therefore it is not a priority. 

PEP8 is a bureaucratic instrument, Python as a language forces to write readable code. There is no need for someone to tell you how much symbols there should be on a line or how to make imports. If you weren't supposed to import on one line, why could you?

**Comment by Aareon** [comment]
You do have a point. I respect that and as to not drive this issue any more off topic, I’ll refrain from bringing it up any more.

Everyone has their own preferred coding styles.

**Comment by EggPool** [comment]
@HarryR 
You're absolutely right and pointed out the core issues.

Like a lot of projects: it's only once it's done that you know how to do it well, and you'd have to reboot from a clean state with your experience alone.
Then it's a question of resources and priorities.

As for myself, there are also lots of things I'd like to improve in the current codebase. But it works, and I have to choose between working on wanted features or refactoring. 

Hard choice. 
The later we refactor, the more we will have to, but the more experience and features we will also have. 
If we refactor sooner, no new feature for a while, and high risk of regression, since we don't have proper tests yet and the real world is likely to smash our face.

That explains my will to modularize the code, piece by piece, not touching the code itself at start but just splitting the code base. So we minimize the regression risk, and have smaller modules to work with. With a single module, we can write extensive tests that validate on the current state of the module, and work on refactoring, but module by module, with tests.

I hope this way we will be able to refactor and add features side by side.

**Comment by hclivess** [comment]
This sounds amazing and I admit to have underestimated the importance of modularity. The original idea was to have everything as simple as possible and within one file, but it appears this was not an ideal choice, because it hinders cooperation and product development monitoring. It also makes things more confusing for experienced developers who have the right habits.

**Comment by Aareon** [comment]
That’s one of the hardest things of such a large project with so many people relying on it. Refactoring and at the same time doing tests on everything so as to not regress like Egg said. I have to say, you guys are much more eloquent than I am.

**Comment by hclivess** [comment]
Closed here for now, work has been done on Google protocol buffer system PoC, will be tracked in Trello
